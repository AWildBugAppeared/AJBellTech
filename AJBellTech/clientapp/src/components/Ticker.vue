<template>
  <v-row justify="center">
    <v-col cols="8">
      <v-data-table
        :headers="headers"
        :items="tickerDataItems"
        :sort-by="['15m']"
      >
        <template #[`item.currency`]="{ item }">
          {{ item.currency.toUpperCase() }}
        </template>
      </v-data-table>
    </v-col>
  </v-row>
</template>

<script lang="ts">
import { mapActions } from "vuex";
import Vue from "vue";

import { ITickerDataItem } from "@/dto-models/ticker-data-item";

export default Vue.extend({
  name: "Ticker",

  data() {
    return {
      headers: [
        { text: "Currency", value: "currency" },
        { text: "15m", value: "15m" },
        { text: "Last", value: "last" },
        { text: "Buy", value: "buy" },
        { text: "Sell", value: "sell" },
        { text: "Symbol", value: "symbol" },
      ],
      tickerDataItems: [] as ITickerDataItem[],
      tickerDataRefreshInterval: 3000,
    };
  },

  created() {
    this.loadTickerData();

    setInterval(() => {
      this.loadTickerData();
    }, this.tickerDataRefreshInterval);
  },

  methods: {
    ...mapActions("ticker", ["getTickerData"]),

    loadTickerData() {
      this.getTickerData().then((tickerDataItems: ITickerDataItem[]) => {
        this.tickerDataItems = tickerDataItems;
      });
    },
  },
});
</script>

<style lang="scss" scoped></style>
