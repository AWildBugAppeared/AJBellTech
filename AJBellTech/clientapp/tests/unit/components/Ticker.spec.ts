/* eslint-disable  @typescript-eslint/no-explicit-any */

import Vue from "vue";
import Vuetify from "vuetify";
import Vuex from "vuex";
import { createLocalVue, mount, Wrapper } from "@vue/test-utils";

import Ticker from "@/components/Ticker.vue";

import { TestTickerData } from "../test-helpers/test-ticker-data";

Vue.use(Vuex);

jest.useFakeTimers();

describe("Ticker.vue", () => {
  const tickerDataItems = TestTickerData.getTickerDataItems();

  const actions = {
    getTickerData: jest.fn().mockResolvedValue(tickerDataItems),
  };

  const localVue = createLocalVue();
  let wrapper: Wrapper<Vue & { [key: string]: any }>;
  let mountWrapper: () => void;

  beforeEach(() => {
    const vuetify = new Vuetify({
      mocks: {
        $vuetify: {},
      },
    });
    Vue.use(Vuetify);
    document.body.setAttribute("data-app", "true");

    mountWrapper = () => {
      const storeData = {
        modules: {
          ticker: {
            namespaced: true,
            actions: actions as any,
          },
        },
      };
      const $store = new Vuex.Store(storeData);

      wrapper = mount(Ticker, {
        localVue,
        vuetify,
        data() {
          return {
            tickerDataRefreshInterval: 3000,
          };
        },
        mocks: {
          $store,
        },
      });
    };
  });

  afterEach(() => {
    wrapper.destroy();
    jest.resetModules();
    jest.clearAllMocks();
  });

  describe("rendering", () => {
    it("should render ticker data ordered by 15m", async () => {
      mountWrapper();
      await wrapper.vm.$nextTick();
      await wrapper.vm.$nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe("created", () => {
    it("should get ticker data on creation", () => {
      const spy = jest.spyOn((Ticker as any).options.methods, "loadTickerData");
      mountWrapper();

      expect(spy.mock.calls.length).toBe(1);
    });

    it("should get ticker data every 3 seconds", () => {
      const spy = jest.spyOn((Ticker as any).options.methods, "loadTickerData");
      mountWrapper();

      for (let i = 1; i < 4; i++) {
        jest.advanceTimersByTime(3000);
        //Since the methods is called outside the timer once, we do i+1
        expect(spy.mock.calls.length).toBe(i + 1);
      }
    });
  });

  describe("methods", () => {
    describe("loadTickerData", () => {
      it("should get ticker data", () => {
        const spy = jest.spyOn(
          (Ticker as any).options.methods,
          "loadTickerData"
        );
        mountWrapper();
        spy.mockRestore();

        wrapper.vm.loadTickerData();

        expect(actions.getTickerData.mock.calls.length).toBe(1);
      });
    });
  });
});
