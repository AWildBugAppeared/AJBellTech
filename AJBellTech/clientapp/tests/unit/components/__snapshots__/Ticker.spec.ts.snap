// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticker.vue rendering should render ticker data ordered by 15m 1`] = `
<div class="row justify-center">
  <div class="col col-8">
    <div class="v-data-table v-data-table--has-bottom theme--light">
      <div class="v-data-table__wrapper">
        <table>
          <colgroup>
            <col class="">
            <col class="">
            <col class="">
            <col class="">
            <col class="">
            <col class="">
          </colgroup>
          <thead class="v-data-table-header">
            <tr>
              <th role="columnheader" scope="col" aria-label="Currency: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Currency</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="15m: Sorted ascending. Activate to sort descending." aria-sort="ascending" class="text-start sortable active asc"><span>15m</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Last: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Last</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Buy: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Buy</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Sell: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Sell</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Symbol: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Symbol</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            </tr>
          </thead>
          <tbody>
            <tr class="">
              <td class="text-start">
                GBP
              </td>
              <td class="text-start">36930.65</td>
              <td class="text-start">36930.65</td>
              <td class="text-start">36930.65</td>
              <td class="text-start">36930.65</td>
              <td class="text-start">£</td>
            </tr>
            <tr class="">
              <td class="text-start">
                EUR
              </td>
              <td class="text-start">42835.52</td>
              <td class="text-start">42835.52</td>
              <td class="text-start">42835.52</td>
              <td class="text-start">42835.52</td>
              <td class="text-start">€</td>
            </tr>
            <tr class="">
              <td class="text-start">
                CHF
              </td>
              <td class="text-start">47516.6</td>
              <td class="text-start">47516.6</td>
              <td class="text-start">47516.6</td>
              <td class="text-start">47516.6</td>
              <td class="text-start">CHF</td>
            </tr>
            <tr class="">
              <td class="text-start">
                USD
              </td>
              <td class="text-start">51094.71</td>
              <td class="text-start">51094.71</td>
              <td class="text-start">51094.71</td>
              <td class="text-start">51094.71</td>
              <td class="text-start">$</td>
            </tr>
            <tr class="">
              <td class="text-start">
                CAD
              </td>
              <td class="text-start">64664.45</td>
              <td class="text-start">64664.45</td>
              <td class="text-start">64664.45</td>
              <td class="text-start">64664.45</td>
              <td class="text-start">$</td>
            </tr>
            <tr class="">
              <td class="text-start">
                AUD
              </td>
              <td class="text-start">66472.79</td>
              <td class="text-start">66472.79</td>
              <td class="text-start">66472.79</td>
              <td class="text-start">66472.79</td>
              <td class="text-start">$</td>
            </tr>
            <tr class="">
              <td class="text-start">
                SGD
              </td>
              <td class="text-start">68589.54</td>
              <td class="text-start">68589.54</td>
              <td class="text-start">68589.54</td>
              <td class="text-start">68589.54</td>
              <td class="text-start">$</td>
            </tr>
            <tr class="">
              <td class="text-start">
                NZD
              </td>
              <td class="text-start">71301.09</td>
              <td class="text-start">71301.09</td>
              <td class="text-start">71301.09</td>
              <td class="text-start">71301.09</td>
              <td class="text-start">$</td>
            </tr>
            <tr class="">
              <td class="text-start">
                PLN
              </td>
              <td class="text-start">196807.89</td>
              <td class="text-start">196807.89</td>
              <td class="text-start">196807.89</td>
              <td class="text-start">196807.89</td>
              <td class="text-start">zł</td>
            </tr>
            <tr class="">
              <td class="text-start">
                BRL
              </td>
              <td class="text-start">290785.12</td>
              <td class="text-start">290785.12</td>
              <td class="text-start">290785.12</td>
              <td class="text-start">290785.12</td>
              <td class="text-start">R$</td>
            </tr>
          </tbody>
        </table>
      </div>
      <div class="v-data-footer">
        <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
            <div class="v-input__control">
              <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-15" class="v-input__slot">
                <div class="v-select__slot">
                  <div class="v-select__selections">
                    <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-15" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                  </div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                  </div><input type="hidden" value="10">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="v-data-footer__pagination">1-10 of 23</div>
        <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
        <div class="v-data-footer__icons-after"><button type="button" class="v-btn v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
      </div>
    </div>
  </div>
</div>
`;
